clientconn.go

```diff
@@ -860,176 +854,267 @@ func (ac *addrConn) errorf(format string, a ...interface{}) {
-			connectDeadline = ac.connectDeadline
-			ac.backoffDeadline = time.Time{}
-			ac.connectDeadline = time.Time{}
-			ac.connectRetryNum = 0
+		if ac.state == connectivity.Shutdown {
+			ac.mu.Unlock()
+			return
 		}
+		if ac.ready != nil {
+			close(ac.ready)
+			ac.ready = nil
+		}
+		ac.transport = nil
+
+		backoffIdx := ac.backoffIdx
+		backoffFor := ac.dopts.bs.Backoff(backoffIdx)
+
+		// This will be the duration that dial gets to finish.
+		dialDuration := getMinConnectTimeout()
+		if backoffFor > dialDuration {
+			// Give dial more time as we keep failing to connect.
+			dialDuration = backoffFor
+		}
+		start := time.Now()
+		connectDeadline := start.Add(dialDuration)
+		ac.backoffDeadline = start.Add(backoffFor)
+		ac.connectDeadline = connectDeadline
+
+		ac.mu.Unlock()
+
+		ac.cc.mu.RLock()
+		ac.dopts.copts.KeepaliveParams = ac.cc.mkp
+		ac.cc.mu.RUnlock()
+
+		ac.mu.Lock()
+
 		if ac.state == connectivity.Shutdown {
 			ac.mu.Unlock()
-			return errConnClosing
+			return
 		}
+
 		ac.printf("connecting")
 		if ac.state != connectivity.Connecting {
 			ac.state = connectivity.Connecting
 			ac.cc.handleSubConnStateChange(ac.acbw, ac.state)
 		}
-		// copy ac.addrs in case of race
-		addrsIter := make([]resolver.Address, len(ac.addrs))
-		copy(addrsIter, ac.addrs)
+
+		addr := ac.addrs[ac.addrIdx]
 		copts := ac.dopts.copts
 		ac.mu.Unlock()
-		connected, err := ac.createTransport(connectRetryNum, ridx, backoffDeadline, connectDeadline, addrsIter, copts)
-		if err != nil {
-			return err
-		}
-		if connected {
-			return nil
+
+		if err := ac.createTransport(backoffIdx, addr, copts, connectDeadline); err != nil {
+			// errReadTimeOut indicates that the handshake was not received before
+			// the deadline. We exit here because the transport's reader goroutine will
+			// use onClose to reset the transport.
+			if err == errReadTimedOut {
+				return
+			}
+
+			continue
 		}
+
+		return
 	}
 }
 
+var errReadTimedOut = errors.New("read timed out")
+
 // createTransport creates a connection to one of the backends in addrs.
-// It returns true if a connection was established.
-func (ac *addrConn) createTransport(connectRetryNum, ridx int, backoffDeadline, connectDeadline time.Time, addrs []resolver.Address, copts transport.ConnectOptions) (bool, error) {
-	for i := ridx; i < len(addrs); i++ {
-		addr := addrs[i]
-		target := transport.TargetInfo{
-			Addr:      addr.Addr,
-			Metadata:  addr.Metadata,
-			Authority: ac.cc.authority,
+func (ac *addrConn) createTransport(backoffNum int, addr resolver.Address, copts transport.ConnectOptions, connectDeadline time.Time) error {
+	timedOutWaitingForHandshake := make(chan struct{})
+	shutdownOccurred := make(chan struct{})
+	allowedToReset := make(chan struct{})
+
+	onGoAway := func(r transport.GoAwayReason) {
+		ac.mu.Lock()
+		ac.adjustParams(r)
+		ac.mu.Unlock()
+		go ac.resetTransport(false)
+	}
+
+	prefaceReceived := make(chan struct{})
+	prefaceReceivedMu := sync.Mutex{}
```

Comment:
To avoid this mutex, you could use a different channel for when onClose was called, and block on `prefaceRecevied` _or_ `closed` when waiting on the preface timer.

Also, I think the way it is now, it's possible to get a double-close.  If `onClose` and `onPrefaceReceipt` are both called at ~the same time and `onClose` wins the lock, then you'll double-close `prefaceReceived`.  Probably best to just have the second channel.

Reply:
Shoot, you're right.

Done.